File: 41-first-missing-positive/41-first-missing-positive.swift
class Solution {
    func firstMissingPositive(_ nums: [Int]) -> Int {
        var arr = nums
        var size = nums.count
        
        for index in 0..<arr.count{
            if arr[index] < 0 || arr[index] > size{
                arr[index] = 0
            }
        }
        
        for index in 0..<arr.count{
            var newIndex = abs(arr[index])
            print(newIndex)
            if newIndex > 0  && newIndex <= arr.count{
                if arr[newIndex - 1] == 0{
                arr[newIndex - 1] = -(arr.count + 1)
                }else if arr[newIndex - 1] > 0 {
                     arr[newIndex - 1] *= -1
                }
            }
            
        }
        print(arr)
        for index in 0..<arr.count{
            if arr[index] >= 0 && arr[index] <= arr.count{
                return index + 1
            }
        }
        
        return arr.count + 1
    }
}

File: 41-first-missing-positive/41-first-missing-positive.cpp
class Solution {
public:
    int firstMissingPositive(vector<int>& nums) {
        int size = nums.size();
        
        int oneCount = 0 ;
        
        
        for(int index = 0; index < size; index ++){
            if(nums[index] == 1) oneCount ++;
            if(nums[index] <= 0 || nums[index] > size) nums[index] = 1;
        }
        
        if(oneCount == 0) return 1;
        int index = 0 ;
        
        while(index < size){
            int j = nums[index] - 1;
            if(nums[index] != nums[j]){
                swap(nums[index], nums[j]);
            }else{
                index ++;
            }

        }
        int ans = size + 1;
        for(int index = 0; index < size; index ++ ){
            if( (nums[index] - 1) != index ){
                return index + 1;
            }
        }
        return ans;
    }
};


